internal/core/block/signature.go

```diff
@@ -97,57 +97,100 @@ func (sig *Signature) GenerateNode() ipld.Node {
 	return bindnode.Wrap(sig, SignatureSchema).Representation()
 }
 
-// VerifyBlockSignature verifies the signature of a block.
+// verifySignature performs the cryptographic verification and returns appropriate results
+func verifySignature(pubKey crypto.PublicKey, signedBytes, sigValue []byte) error {
+	valid, err := pubKey.Verify(signedBytes, sigValue)
+	if err != nil {
+		return err
+	}
+	if !valid {
+		return crypto.ErrSignatureVerification
+	}
+	return nil
+}
+
+// VerifyBlockSignature verifies the signature of a block using the link system.
 // The first return value is true if it actually ran cryptographic verification.
 func VerifyBlockSignature(block *Block, lsys *linking.LinkSystem) (bool, error) {
 	if block.Signature == nil {
 		return false, nil
 	}
 
+	signedBytes, sigBlock, err := getSignedDataAndSignature(block, lsys)
+	if err != nil {
+		return false, err
+	}
+
+	pubKey, err := getPublicKeyFromSignature(sigBlock)
+	if err != nil {
+		return false, err
+	}
+
+	return true, verifySignature(pubKey, signedBytes, sigBlock.Value)
+}
+
+// VerifyBlockSignatureWithKey verifies the signature of a block using a public key.
+// The first return value is true if it actually ran cryptographic verification.
+func VerifyBlockSignatureWithKey(block *Block, lsys *linking.LinkSystem, pubKey crypto.PublicKey) (bool, error) {
+	if block.Signature == nil {
+		return false, nil
+	}
+
+	signedBytes, sigBlock, err := getSignedDataAndSignature(block, lsys)
+	if err != nil {
+		return false, err
+	}
+
+	// Verify that the identity matches the signature's identity
+	if string(sigBlock.Header.Identity) != pubKey.String() {
+		return false, ErrSignaturePubKeyMismatch
+	}
+
+	return true, verifySignature(pubKey, signedBytes, sigBlock.Value)
+}
+
+func getSignedDataAndSignature(block *Block, lsys *linking.LinkSystem) ([]byte, *Signature, error) {
```

Comment:
thought: This is kind of an odd function, as it appears to return two unrelated properties - `signedBytes` is not used to get `sigBlock` and just gets returned untouched by the other lines in this function.  To me this suggests that this function should be broken up and should not deal with `signedBytes`.